crypto/encode_decode/encoder_pkey.c

```diff
@@ -101,6 +102,9 @@ static void collect_encoder(OSSL_ENCODER *encoder, void *arg)
         const OSSL_PROVIDER *prov = OSSL_ENCODER_get0_provider(encoder);
         void *provctx = OSSL_PROVIDER_get0_provider_ctx(prov);
 
+        if (!!(data->keymgmt_prov == prov) != !!data->flag_find_same_provider)
```

Comment:
Both `!!` are unnecessary and require a double take.

I suspect the logic is wrong: if !data->flag_find_same_provider, why would the provider matter?

If the logic is correct, you need to put a comment explaining it in the code.

Reply:
> Both `!!` are unnecessary and require a double take.

True.

> I suspect the logic is wrong: if !data->flag_find_same_provider, why would the provider matter?

If you look further down, you'll see that `OSSL_ENCODER_do_all_provided` is called twice, once with `flag_find_same_provider` set, and once with it unset.  It's to ensure that the encoders where the encoders from the same provider as the keymgmt are considered first.  This avoids problems with encoders not wanting to import just about anything, or keymgmts not willing to export the key data (would be typical for HSMs), and generally avoids having to do the export/import dance unless really necessary.

Reply:
I guess I have my comment right there :wink:

Reply:
That's what I was typing...

Reply:
Comment added

Reply:
... and `!!` removed